File: app/src/main/java/me/devsaki/hentoid/fragments/intro/PermissionIntroFragment.java
package me.devsaki.hentoid.fragments.intro;

import androidx.fragment.app.Fragment;

public class PermissionIntroFragment extends Fragment /*implements SlidePolicy*/ {

    /*
    private IntroActivity_ parentActivity;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof IntroActivity_) {
            parentActivity = (IntroActivity_) context;
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.intro_slide_02, container, false);
    }

    @Override
    public boolean isPolicyRespected() {
        Context context = getContext();
        if (context == null) return false;
        int permissionCode = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        return permissionCode == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onUserIllegallyRequestedNextPage() {
        invokeAskPermission();
    }

    private void invokeAskPermission() {
        if (parentActivity != null)
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, PermissionHelper.RQST_STORAGE_PERMISSION);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode != PermissionHelper.RQST_STORAGE_PERMISSION) return;
        if (permissions.length == 0) return;
        if (!permissions[0].equals(Manifest.permission.WRITE_EXTERNAL_STORAGE)) return;
        if (grantResults.length == 0) return;
        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            parentActivity.onPermissionGranted();
        } else if (shouldShowRequestPermissionRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            showPermissionComplaint();
        } else {
            showPermissionComplaintManual();
        }
    }

    private void showPermissionComplaint() {
        View view = getView();
        if (view == null) return;
        Snackbar.make(view, R.string.permissioncomplaint_snackbar, LENGTH_LONG)
                .setAction(android.R.string.ok, v -> invokeAskPermission())
                .show();
    }

    private void showPermissionComplaintManual() {
        View view = getView();
        if (view == null) return;
        Snackbar.make(view, R.string.permissioncomplaint_snackbar_manual, LENGTH_LONG)
                .setAction(android.R.string.ok, v -> invokeOpenSettings())
                .show();
    }

    private void invokeOpenSettings() {
        Uri uri = Uri.parse("package:" + BuildConfig.APPLICATION_ID);
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, uri);
        startActivity(intent);
    }

     */
}

File: app/src/main/java/me/devsaki/hentoid/fragments/intro/PermissionIntroFragment_.kt
package me.devsaki.hentoid.fragments.intro

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.github.appintro.SlidePolicy
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import me.devsaki.hentoid.R
import me.devsaki.hentoid.activities.IntroActivity_
import me.devsaki.hentoid.databinding.IntroSlide02Binding
import me.devsaki.hentoid.util.Preferences

class PermissionIntroFragment_ :
    Fragment(R.layout.intro_slide_02), SlidePolicy {

    private lateinit var parentActivity: IntroActivity_
    private var _binding: IntroSlide02Binding? = null
    private val binding get() = _binding!!

    // Ask for permissions
    private val requestPermissionLauncher =
        registerForActivityResult(
            ActivityResultContracts.RequestPermission()
        ) { isGranted: Boolean ->
            if (isGranted) {
                lifecycleScope.launch {
                    delay(75)
                    parentActivity.nextStep()
                }
            } else {
                binding.modeSelect.check(R.id.mode_browser)
            }
        }


    // If user should be allowed to leave this slide
    override val isPolicyRespected: Boolean
        get() {
            if (R.id.mode_browser == binding.modeSelect.checkedButtonId) return true
            val permissionCode =
                ContextCompat.checkSelfPermission(
                    requireContext(),
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
                )
            return (permissionCode == PackageManager.PERMISSION_GRANTED)
        }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is IntroActivity_) {
            parentActivity = context
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = IntroSlide02Binding.inflate(inflater, container, false)

        binding.modeSelect.addOnButtonCheckedListener { _, _, b ->
            if (!b) return@addOnButtonCheckedListener

            Preferences.setBrowserMode(binding.modeBrowser.isChecked)
            binding.descTxt.setText(if (binding.modeBrowser.isChecked) R.string.slide_02_browser_mode_description else R.string.slide_02_library_mode_description)
        }

        val browserMode = Preferences.getBrowserMode()
        if (browserMode != null) binding.modeSelect.check(if (browserMode) R.id.mode_browser else R.id.mode_library)

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onUserIllegallyRequestedNextPage() {
        when (PackageManager.PERMISSION_GRANTED) {
            ContextCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            ) -> {
                parentActivity.nextStep()
            }
            else -> {
                requestPermissionLauncher.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
            }
        }
    }
}